File: internal/daemon/handlers/start.go
package handlers

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/Jaro-c/Lynx/internal/daemon/manager"
	"github.com/Jaro-c/Lynx/internal/ipc/protocol"
	"github.com/Jaro-c/Lynx/internal/ipc/transport"
	"github.com/Jaro-c/Lynx/internal/jsonx"
)

// StartHandler handles the start command.
func StartHandler(mgr *manager.Manager, privileged bool) transport.CommandHandler {
	return func(ctx context.Context, params jsonx.RawMessage) (jsonx.RawMessage, error) {
		var req protocol.StartRequest
		if err := jsonx.Unmarshal(params, &req); err != nil {
			return nil, fmt.Errorf("ERR_BAD_REQUEST: %w", err)
		}

		spec := req.Spec
		if spec.ID == "" {
			return nil, errors.New("ERR_BAD_REQUEST: spec ID is required")
		}

		// Peer identity (uid/gid/pid) comes from SO_PEERCRED — the IPC
		// server attaches it to ctx for every request. Future per-user
		// isolation modes (explicit_user, app_user) will consult it.
		identity, ok := ctx.Value(transport.ContextKeyIdentity).(*transport.Identity)
		if !ok {
			return nil, errors.New("INTERNAL_ERROR: identity not found")
		}

		// Start process via Daemon logic
		procInfo, err := StartProcess(mgr, spec, identity, privileged)
		if err != nil {
			return nil, err
		}

		respData := protocol.StartResponseData{
			ProcID:    procInfo.ID,
			PID:       procInfo.PID,
			Status:    string(procInfo.State),
			CreatedAt: time.Now().Format(time.RFC3339),
		}

		return jsonx.Marshal(respData)
	}
}

File: internal/daemon/handlers/start.rs
//! The `start` verb — the daemon's front door.
//!
//! Mirrors `internal/daemon/handlers/start.go`. Parses the
//! [`StartRequest`](crate::ipc::protocol::StartRequest) envelope, hands
//! the spec to [`start_process`](super::service::start_process), and
//! packages the response as a [`StartResponseData`].
//!
//! Peer identity comes from the [`RequestContext`] the dispatcher attaches
//! to every request — `SO_PEERCRED` populated it before the handler ran.

use crate::daemon::handlers::service::{start_process, SharedManager};
use crate::ipc::protocol::{RawMessage, StartRequest, StartResponseData};
use crate::ipc::transport::RequestContext;
use crate::jsonx;

/// Build the `start` handler. `privileged` is the daemon's own mode —
/// `true` for the system instance (root or `unitpm`), `false` for a
/// user-mode daemon. The `policy::authorize_start` check enforces the
/// security boundary.
pub fn start_handler(
	mgr: SharedManager,
	privileged: bool,
) -> impl Fn(RequestContext, RawMessage) -> Result<RawMessage, String> + Send + Sync + 'static {
	move |ctx, params| {
		let req: StartRequest =
			jsonx::unmarshal(params.as_bytes()).map_err(|e| format!("ERR_BAD_REQUEST: {e}"))?;

		if req.spec.id.is_empty() {
			return Err("ERR_BAD_REQUEST: spec ID is required".into());
		}

		let info = start_process(&mgr, req.spec, &ctx.identity, privileged)?;

		let data = StartResponseData {
			id: info.id.clone(),
			proc_id: Some(info.id),
			pid: Some(info.pid as i32),
			status: Some(info.state.as_str().to_string()),
			message: None,
			created_at: None,
		};

		jsonx::marshal(&data)
			.map(|b| RawMessage::from_bytes(&b))
			.map_err(|e| e.to_string())
	}
}
